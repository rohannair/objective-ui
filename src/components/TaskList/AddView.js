import React, { Component, PropTypes } from 'react'
import styled from 'styled-components'

import Checkbox from '../Forms/Checkbox'
import TextInput from '../Forms/TextInput'
import { StyledButton as Button } from '../Button'

class AddView extends Component {
  constructor(props) {
    super(props)

    this.defaultTaskState = {
      id: 0,
      title: '',
      isComplete: false
    }

    this.state = props.task
    ? props.task
    : this.defaultTaskState
  }

  render() {
    const { title, isComplete} = this.state

    return (
      <div className={this.props.className}>
        <Checkbox
          isChecked={isComplete}
          onChange={this._updateTask('isComplete')}
          className="checkbox"
        />

        <TextInput
          defaultValue={title}
          type="text"
          onChange={this._updateTask('title')}
          className="title"
        />

        <div className={'actions'}>
          <Button link icon onClick={this._handleSave}>
            <i className="zmdi zmdi-save" />
          </Button>

          <Button cancel link icon onClick={this.props.onCancel}>
            &times;
          </Button>
        </div>
      </div>
    )
  }

  _handleKeyPress = (event) => {
    if (event.key === 'Enter') {
      this._handleSave()
    }
  }

  _handleSave = () => {
    const task = this.state
    if (task.title === '') return

    this.setState(this.defaultTaskState)
    this.props.saveTask(task)
  }

  _updateTask = (name) => val => {
    this.setState(prev => ({
      ...prev,
      [name]: val
    }))
  }
}

AddView.propTypes = {
  onCancel: PropTypes.func,
  saveTask: PropTypes.func.isRequired
}

AddView.defaultProps = {
  onCancel: () => {}
}

const StyledAddView = styled(AddView)`
  display: flex;
  flex: nowrap;
  justify-content: space-around;
  align-items: baseline;
  align-content: center;
  width: 100%;

  .checkbox {
    flex: 0 0 auto;
    margin: 0;
  }

  .title {
    flex: 1 0 auto;
    padding: 0 0 0 0.625rem;

    & input {
      padding: 0;
    }
  }

  .actions {
    display: flex;

  }

`

export default StyledAddView
