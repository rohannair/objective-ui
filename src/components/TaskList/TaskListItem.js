import React from 'react'
import styled from 'styled-components'
import classNames from 'classnames'

import Checkbox from '../Forms/Checkbox'
import { StyledButton as Button } from '../Button'

const TaskListItem = styled(({
  task,
  completeTask,
  className
}) => {
  const checkboxStyle = classNames({
    completed: task.isComplete,
    checkbox: true
  })

  return (
    <div className={className}>
      <Checkbox
        isChecked={task.isComplete}
        label={task.title}
        onChange={completeTask}
        className={checkboxStyle}
      />
      <Button className={'edit'} icon><i className="zmdi zmdi-edit" /></Button>
    </div>
  )
})`
  display: flex;
  margin: 2px 0;

  .checkbox {
    flex: 1 0 auto;

    &.completed span {
      color: #666;
      text-decoration: line-through;
    }
  }

  .edit {
    flex: 0 0 auto;
    visibility: hidden;
  }

  &:hover .edit {
    visibility: visible;
  }
`
export default TaskListItem
