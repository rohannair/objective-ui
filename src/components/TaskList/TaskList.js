import React, { Component, PropTypes } from 'react'
import styled from 'styled-components'
import gql from 'graphql-tag'

import Button, { StyledButton } from '../Button'
import TaskListItem from './TaskListItem'
import AddView from './AddView'

const TaskList = styled(({
  tasks,
  saveTask,
  editTask,
  className,
  isCollaborator
}) => {
  const handleCompleteTask = (task) => v => (isCollaborator ? editTask({...task, isComplete: v}) : null)
  const handleEditTask = (task) => v => (isCollaborator ? editTask({...task, title: v}) : null)

  const taskListBody = tasks && tasks.length
    ? tasks.map(t => <TaskListItem task={t} key={t.id} completeTask={handleCompleteTask(t)} editTask={handleEditTask(t)} className='task' />)
    : <div>No tasks :( Add some!</div>

  return (
    <div className={className}>
      <h3>Task List:</h3>

      <div className={'list'}>
        { taskListBody }

        { isCollaborator ? <AddTaskListItem saveTask={saveTask} className='task' /> : null }
      </div>
    </div>
  )
})`
  .list {
    margin: 20px 0;
  }

  .task {
    margin: 2px 0;
    padding: 7px 3px;
  }
`

class AddTaskListItem extends Component {
  constructor(props) {
    super(props)

    this.state = {
      isAdding: false
    }
  }

  render() {
    return !this.state.isAdding
      ? <Button inline onClick={this._toggleView}>+ Add Task</Button>
      : <AddView saveTask={this._handleSave} onCancel={this._toggleView} className={this.props.className} />
  }

  _toggleView = () => {
    this.setState(prev => ({
      isAdding: !prev.isAdding
    }))
  }

  _handleSave = (task) => {
    this._toggleView()
    this.props.saveTask(task)
  }
}

const CREATE_TASK = gql`
  mutation createTask($title: String!, $isComplete: Boolean!, $objective: String!) {
    createTask(title: $title, isComplete: $isComplete, objective: $objective) {
      id
      title
      isComplete
    }
  }
`

const EDIT_TASK = gql`
  mutation editTask($id: String!) {
    editTask(id: $id) {
      id
      title
      isComplete
    }
  }
`

TaskList.mutations = {
  CREATE_TASK,
  EDIT_TASK
}

export default TaskList
