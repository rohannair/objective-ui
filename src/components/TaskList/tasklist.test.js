import React from 'react'

import TaskList from './index.js'

describe('Test suite for TaskList component', () => {
  it('TaskList should exist', () => {
    let wrapper = shallow(<TaskList/>)
    expect(wrapper).toMatchSnapshot()
  })
})

