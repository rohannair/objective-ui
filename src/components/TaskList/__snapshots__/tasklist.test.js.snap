exports[`Test suite for TaskList component TaskList should exist 1`] = `
<Component
  className="dlvdBO" />
`;
